src/state.rs

```diff
@@ -0,0 +1,214 @@
+}
+
+/// The `State<I>` struct maintains the current position in the stream `I` using
+/// the `Positioner` trait to track the position.
+///
+/// ```
+/// # extern crate combine;
+/// # use combine::{token, Parser, ParseError};
+/// # use combine::primitives::{Error};
+/// # use combine::state::{State, IndexPositioner};
+/// # fn main() {
+///     let result = token('9')
+///         .message("Not a nine")
+///         .parse(State::new("8", IndexPositioner::<char, &str>::new()));
+///     assert_eq!(result, Err(ParseError {
+///         position: 0,
+///         errors: vec![
+///             Error::Unexpected('8'.into()),
+///             Error::Expected('9'.into()),
+///             Error::Message("Not a nine".into())
+///         ]
+///     }));
+/// # }
+/// ```
+#[derive(Clone, PartialEq)]
+pub struct State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position>,
+          I::Position: Clone + Ord
+{
+    /// The input stream used when items are requested
+    pub input: I,
+    /// The positioner used to update the current position
+    pub positioner: X,
+}
+
+impl<I, X> fmt::Debug for State<I, X>
+    where I: Stream + fmt::Debug,
+          X: Positioner<Position = I::Position>,
+          I::Position: Clone + Ord + fmt::Debug
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f,
+               "State {{ position: {:?}, input: {:?} }}",
+               self.positioner.position(),
+               self.input)
+    }
+}
+
+impl<I, X> State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position>,
+          I::Position: Clone + Ord
+{
+    /// Creates a new `State<I, X>` from an input stream and a positioner.
+    pub fn new(input: I, positioner: X) -> State<I, X> {
+        State {
+            input: input,
+            positioner: positioner,
+        }
+    }
+}
+
+impl<I, X> StreamOnce for State<I, X>
+    where I: Stream,
+          X: Positioner<Position = I::Position, Item = I::Item>,
+          I::Position: Clone + Ord
+{
+    type Item = I::Item;
+    type Range = I::Range;
+    type Position = I::Position;
+
+    #[inline]
+    fn uncons(&mut self) -> Result<I::Item, Error<I::Item, I::Range>> {
+        match self.input.uncons() {
+            Ok(c) => {
+                self.positioner.update(&c);
+                Ok(c)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    #[inline(always)]
+    fn position(&self) -> Self::Position {
+        self.positioner.position()
+    }
+}
+
+/// The `IndexPositioner<Item, Range>` struct maintains the current index into the stream `I`.  The
+/// initial index is index 0.  Each `Item` consumed increments the index by 1; each `range` consumed
+/// increments the position by `range.len()`.
+#[derive(Clone, PartialEq)]
+pub struct IndexPositioner<Item, Range>(usize,
+                                        PhantomData<Item>,
+                                        PhantomData<Range>);
```

Comment:
This might not matter but write `PhantomData<fn () -> (Item, Range)>` which will make covariance/contravariance more correct and makes it so only one `PhantomData` is needed.